source/extensions/clusters/eds/eds.cc

```diff
@@ -378,6 +419,27 @@ bool EdsClusterImpl::updateHostsPerLocality(
 void EdsClusterImpl::onConfigUpdateFailed(Envoy::Config::ConfigUpdateFailureReason reason,
                                           const EnvoyException*) {
   ASSERT(Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure != reason);
+  // Config failure may happen if Envoy times out waiting for the EDS resource.
+  // If it is a timeout, the eds resources cache is enabled,
+  // and there is a cached ClusterLoadAssignment, then the cached assignment should be used.
+  if (reason == Envoy::Config::ConfigUpdateFailureReason::FetchTimedout &&
```

Comment:
Question: what is the `initial_fetch_timeout` typically specified to?  The default seems to be 15s.  This means that if the management server wants to save on bandwidth/data exchange, then it won't send the EDS assignment updates if they haven't changed, but if the `intial_fetch_timeout` is long, then the Envoy client will just be hanging until the timeout before it can warm the cluster.  That could result in an abnormally long delay.  Am I understanding this trade-off correctly?

If so, what about the alternative of always warming with cached assignments, if available, and if the management server sends assignment updates after that, the cluster can just be updated with those assignments and proceed.

Reply:
That's a good question, and I tried to give more context in the original issue (#26749).

The `initial_fetch_timeout` tells Envoy how long to wait before changing a resource from "warming" state to "active". One thing to note about Envoy's implementation is that this value is used not only for the "initial", but also for updates.

The problem is for coupled resources that are meant to be changed "together" but the protocol doesn't really allow for it. For example, an EDS cluster and its assignment may need to be updated together.
This [test](https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/test/integration/ads_integration_test.cc#L560-L613) for example, was introduced to ensure that a non-TLS cluster that has been modified to use TLS, will only be active after its assignment was fetched. this is how Envoy works at the moment.

Note that this is a place where Envoy's implementation doesn't follow the xDS-protocol. The protocol says that the client should use the cached object (if it has it) immediately. However, this will break current use-cases of Envoy, like the ones that do non-TLS -> TLS upgrades.

The current solution is somewhat in-between: use the cached object after a timeout. In case an "upgrade" occurs, Envoy will wait for the assignment before it makes the cluster active. If no update comes, Envoy will "eventually" (timeout) use the cached object.

LMK if you'd like more context here, or if you think of a better approach.

Reply:
> Thanks for that explanation Adi. That clarifies it a lot.
> 
> My concern with using the `initial_fetch_timeout` is that you could have a cluster remain in the warmed state for quite a while (~15s if default values are used). I suppose this would occur if the xDS management server decides to _not_ send EDS assignment updates if they haven't changed.
> 
Or if the xDS management server went down before sending the EDS response. In case the assignment hasn't changed, the updated cluster will now use the cached assignment (rather than an empty assignment previously).

> IIUC, the non-TLS => TLS cluster update use case breaks if we immediately use cached EDS assignments b/c the cached assignments may be pointing to endpoints that don't support the required certificates for TLS? Or is it because the active connections will be maintained without new assignments, so we'll be sending data over non-TLs connections? Or is it something else?
> 
Yes. Essentially it is expected that the cluster will need a different set of endpoints.

> Just a couple (probably not-well-thought-out) ideas for your consideration:
> 
> 1. Should a different, new value be used for the timeout for which the default timeout is shorter (say, 2s?)?
> 2. Or should we change the `initial_fetch_timeout` default from 15s to something shorter?  Looks like we can do that in code here: https://github.com/envoyproxy/envoy/blob/7fb3f62b9edceacbce37c90b1e1e26b06abbe370/source/common/config/utility.cc#L198
>     (note: I sent [config: Update the comment on initial_fetch_timeout default value #28930](https://github.com/envoyproxy/envoy/pull/28930) to correct the comment on the default value in the header file).

I'll reach you offline to discuss this, and maybe come up with other ideas.